### DISCUSSION TITLE
Specifying startup arguments for GHA service containers

### DISCUSSION THREAD

**Discussion by joecotton-wk** [community/community#52675] [Actions]
### Select Topic Area

General

### Body

Hello,

I have a workflow that requires custom arguments to be passed to a service container when it starts up. Is this possible in GHA? Specifically, I'm looking for something that corresponds to the `docker create [OPTIONS] IMAGE [COMMAND] [ARG...]` form of docker create (see [the docs](https://docs.docker.com/engine/reference/commandline/create/)).

I'd expect it to look something like this:

```yaml
services:
  my-test-container:
    image: company/image:tag
    command: binary-name
    arguments: arg1 arg2
```

which would be equivalent to the following: `docker create company/image:tag binary-name arg1 arg2`

I've searched the internet, and all I've found are these two Stack Overflow posts:

- https://stackoverflow.com/questions/60849745/how-can-i-run-a-command-in-github-action-service-containers
- https://stackoverflow.com/questions/64031598/creating-a-minios3-container-inside-a-github-actions-yml-file

Both state that this isn't supported, but neither of them cite an official source. I'd like to know:

- Is this possible
- If not:
   - Would this feature be supported sometime in the future?
   - Are there any workarounds other than the ones mentioned in the SO posts linked above? None of those workarounds are desirable.

**Comment by dgholz** [comment]
It was answered in 2020 on https://github.com/orgs/community/discussions/26688#discussioncomment-3252877 by a engineer was was on the Code to Cloud team at the time, it's not possible.

@chrispat was tagged then, any updates in the meantime if this would be supported in the future?

If I understand it correctly, the workarounds on those SO posts are:
* start the Docker container inside a run step in your job, & get the networking right
* publish a new Docker image based on the original, with your arguments baked in, and use it to run the service in your job

I think those are the only reasonable possibilities.

We're getting around it with the service we want to use, because it can load startup options from a config file. So we:
* bake a VM image with the config file present on disk
* launch our own GitHub runners from that image
* configure our jobs to run on those custom runners
* and use the `jobs.*.services.*.volumes` option to mount the on-disk config into the service container

Which feels like a bit much.

**Comment by jshields** [comment]
This seems like a basic option that GitHub Actions needs to provide. As the author of the question stated, the arguments to the `docker create` command that GHA invokes normally includes "COMMAND" and its own arguments after the image name is specified.
https://docs.docker.com/reference/cli/docker/container/create/

CircleCI allows you to easily specify the command that goes after the image name with the `command` yaml key as the author expected. Both `command` and `entrypoint` can be configured in CircleCI:
https://circleci.com/docs/configuration-reference/#executor-job

It seems quite silly that end users are needing to build their own custom images still after years, because GitHub Actions workflow syntax does not support a `command` yaml key separate from the `options` and `image` that are each their own arguments to the `docker create` command being run. I guess this might be just an issue with `services` containers, but I still think this is a basic feature that needs to be provided.

**Comment by brbrowning21** [comment]
One use case: I want to be able to pass in `--default-authentication-plugin=mysql_native_password` when I run the `mysql:8` service container.

**Comment by salvadoriume** [comment]
I have the same problem. I need to run openfga server as a service.
openfga image itself doesn't run server, it needs command "run" but this isn't available in github service container.
Is there any strong technical problems that it cannot be implemented?
